Switch return type of main from 'void' to 'int'

keith-packard opened the pull request:
As both C and C++ standards require applications running under an OS to
return 'int', adapt that for Zephyr to align with those standard. This also
eliminates errors when building with clang when not using -ffreestanding,
and reduces the need for compiler flags to silence warnings for both clang
and gcc.

---

Related discussion: https://github.com/zephyrproject-rtos/zephyr/issues/54336#issuecomment-1421528758

keith-packard committed:
doc: Switch main return type from void to int.

As both C and C++ standards require applications running under an OS to
return 'int', adapt that for Zephyr to align with those standard. This also
eliminates errors when building with clang when not using -ffreestanding,
and reduces the need for compiler flags to silence warnings for both clang
and gcc.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
samples, tests, boards: Switch main return type from void to int

As both C and C++ standards require applications running under an OS to
return 'int', adapt that for Zephyr to align with those standard. This also
eliminates errors when building with clang when not using -ffreestanding,
and reduces the need for compiler flags to silence warnings for both clang
and gcc.

Most of these changes were automated using coccinelle with the following
script:

@@
@@
- void
+ int
main(...) {
	...
-	return;
+	return 0;
	...
}

Approximately 40 files had to be edited by hand as coccinelle was unable to
fix them.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
compilers: Remove -Wno-main and -Wno-main-return-type compiler flags

These flags were added to avoid warnings when main was declared to return
void. Now that main returns int, those warnings will flag errors.

Signed-off-by: Keith Packard <keithp@keithp.com>

stephanosio commented:
FYI @andyross @peter-mitsis @teburd @nashif @cfriedt

cfriedt reviewed (dismissed):
Just to clarify, I think this should go in after the release, but 100% agree. Thanks!

keith-packard commented:
> It makes sense, arguably a breaking API change though and needs to go through that? What happens with all the existing applications using void main(void)?

Yeah, it's "kinda" an API change. `main` isn't declared in any Zephyr header. If we don't adjust the compiler flags, then there's actually no impact on applications at all. The goal is to allow compiling using clang without `-ffreestanding`, and that requires that main return int.

keith-packard commented:
> Just to clarify, I think this should go in after the release, but 100% agree. Thanks!

Glad to see @stephanosio adjusted the milestone to 3.4.0 -- I wasn't sure what label to apply to make sure this wasn't considered for this upcoming release.

keith-packard committed:
kernel: Switch main return type from void to int

As both C and C++ standards require applications running under an OS to
return 'int', adapt that for Zephyr to align with those standard. This also
eliminates errors when building with clang when not using -ffreestanding,
and reduces the need for compiler flags to silence warnings for both clang
and gcc

Signed-off-by: Keith Packard <keithp@keithp.com>

de-nordic reviewed (dismissed):
OK for smp_svr sample.

nordicjm reviewed (dismissed):
Will need a release note entry added and needs to be brought up at the architecture working group but change itself is good

Thalley commented on samples/bluetooth/beacon/src/main.c:
```diff
@@ -85,4 +85,5 @@ void main(void)
 	if (err) {
 		printk("Bluetooth init failed (err %d)\n", err);
 	}
+	return 0;
```
One comment that may only apply for BT, but for the BT subsystem we usually keep an empty line after an ending `}`, so this would be

```suggestion
	}
	
	return 0;
```

keith-packard replied:
These changes were done using coccinelle; I see a bunch of value in not messing with the results of that where possible as the tool is less likely to generate mistakes than manual editing? Perhaps a manual clean up pass in the series which adjusts formatting could be done though?

Thalley replied:
> These changes were done using coccinelle; 

It does not have an option to specify empty line after ending bracets?

Thalley commented on samples/bluetooth/broadcast_audio_sink/src/main.c:
```diff
@@ -202,15 +202,15 @@ static void reset(void)
 	if (err) {
 		printk("Init failed (err %d)\n", err);
-		return;
+		return 0;
```
Wouldn't it make more sense to return the error? 
```suggestion
		return err;
```

This applies many places

keith-packard replied:
It might, but Zephyr explicitly ignores the return value, so it really doesn't matter in practice.

Thalley replied:
If we ignore the return value, then why even bother adding the return type `int`? 

If we run these samples as Linux applications or part of CI, we definitely don't want to ignore the return value of the `main` function, as we want to catch errors. 

IMO blindly returning `0` in all cases is incorrect.

keith-packard replied:
Totally aligned with your remarks; it seems like an arbitrary change that has zero benefit and the potential for confusion. I've tried to clarify some of this in the doc updates; perhaps some additional wording would be helpful?

The reason for the switch is because the C and C++ standards *require* that `main` be defined as either  `int main(void)` or `int main(int argc, char **argv)` for "hosted" environments. Right now, the GCC and Clang compiler configurations include special compiler flags (`-Wno-main` and `-Wno-main-return-type`) to suppress the compiler warnings. In the future, when the -ffreestanding flag gets removed, the Clang warning turns into a Clang fatal error.

Feel free to complain to the ISO C and C++ committees about their standards, or to the Clang and GCC compiler writers for enforcing them.

Alternatively, maybe we should change the Zephyr OS so that the return value from main has some meaning?

Thalley replied:
@keith-packard I guess it will be too much effort for you to change the `return 0` to `return err` where applicable? 

> Alternatively, maybe we should change the Zephyr OS so that the return value from main has some meaning?

As mentioned above, it does have some meaning already in certain cases

stephanosio replied:
> If we ignore the return value, then why even bother adding the return type int?

Standard conformance -- some compilers refuse to compile `void main(void)` because that is not one of the allowed types for the `main` function defined by the C standard.

> Alternatively, maybe we should change the Zephyr OS so that the return value from main has some meaning?

The return value could be used by the Zephyr kernel to do something useful (e.g. reboot if the main function returns a non-zero value with the assumption that the application initialisation failed); but, that is beyond the scope of this PR and can be addressed separately if we deem something like that to be necessary in the future.

> If we run these samples as Linux applications or part of CI, we definitely don't want to ignore the return value of the `main` function, as we want to catch errors.

But can we currently run these samples as Linux applications? If not, I do not see how that argument is valid at this time -- the return values can be updated accordingly when the sample is updated to be able to run as a Linux application, or when the Zephyr kernel is updated to actually make use of the return value -- what Zephyr kernel will do with the return value is still an open question and beyond the scope of this PR as described above.

Thalley replied:
> But can we currently run these samples as Linux applications?

Yes :) In Nordic we do in fact also run the samples as part of CI, but given that we haven't "needed" the return values yet (we just parse the stdout data), it's a fair argument to not add those in this PR. 

If we don't add the (proper) return values in this PR, I'll just add them later for the Bluetooth samples myself :)

nandojve commented on samples/subsys/mgmt/updatehub/src/main.c:
```diff
@@ -95,15 +95,15 @@ void main(void)
 		LOG_ERR("Failed to register server certificate");
-		return;
+		return 0;
 	}
 
 	if (tls_credential_add(CA_CERTIFICATE_TAG,
 			       TLS_CREDENTIAL_PRIVATE_KEY,
 			       private_key,
 			       sizeof(private_key)) < 0) {
 		LOG_ERR("Failed to register private key");
-		return;
+		return 0;
```
Hi @keith-packard ,

I noted that these two init errors not return a value. Should those return a -errno code ?

keith-packard replied:
Zephyr explicitly ignores the return value from main(). Maybe the documentation patch should make it explicit that a return value of zero is expected?

nandojve replied:
Since Zephyr don't care about the return value you can ignore my previous comment. Documentation may be good to avoid same questions on the communication channels.

keith-packard replied:
good plan. I've added a sentence to both the c and c++ documents indicating that Zephyr ignores the return value from main and that applications should return zero.

peter-mitsis commented:
Aside from the occasional questionable return value (which others have already pointed out), I do not have any fundamental or design issues with what has been proposed.

gregshue commented on doc/develop/languages/c/index.rst:
```diff
@@ -12,6 +12,11 @@ available as part of the C header files under the :file:`include` directory, so
+The ``main()`` function must have the return type of ``int`` as Zephyr
+applications run in a "hosted" environment as defined by the C
+standard. Zephyr ignores the return value from main, so applications should
+not return status information and should, instead, return zero.
```
Please describe the API policy for `main()` rather than describe current Zephyr implementation ("Zephyr ignores the return value of main").  Something like the statement below would work well.  Ideally this would be documented as a public API (even if the actual declaration was conditionally-compiled out):

- All non-zero return values from `main()` are reserved.
- All meanings of return values from `main()` are reserved.

keith-packard replied:
Yeah, you're right, the docs should state the policy not just the current implementation. Once that policy has been agreed upon, we can update the docs to match as part of this PR.

keith-packard replied:
I've changed the wording to require a return value of zero from main (at least for now).

gregshue commented:
Regarding return value usage and meaning:

Bootloaders are a ubiquitous example of an application needing to cleanly shut down the execution environment before handing the HW and CPU(s) off to another executable. To maintain a composable, scalable, and extensible architecture this clean shut down maps to a reverse `SYS_INIT()` which should be executed by the same thread that called `SYS_INIT()`. Thus, returning from `main()` can be very useful to trigger for a shutdown.

AFAICT, the shutdown sequence either succeeds or should cause a catastrophic failure (i.e., `k_panic()`/`z_except_reason()`). IF true, then this sequence has no need for the return value - returning from main() is sufficient to trigger the shutdown sequence. Shutdown parameters (e.g., the pointer to jump to upon shutdown, whether or not to reset upon shutdown, etc.) may be specific to the subsystems involved so shouldn't expect to be encodable in an `int` or passed via the return value.

It may be wise to leave meaning and usage of non-zero return values for `native_posix`, `unit_testing`, and/or emulation environments rather than let the Zephyr kernel try to take advantage of it.

galak commented:
If we merge this, we need to "revert" (and similar changes in other files for `cmake/compiler/arcmwdt/compiler_flags.cmake`):

commit 0f692f5b84db8ef61f1e85c6ef25278966dc3329
Author: Keith Packard <keithp@keithp.com>
Date:   Mon Feb 6 21:10:29 2023 -0800

    compiler/clang: ignore main-return-type warning for clang

mbolivar-nordic commented:
Arch WG:

- This is a treewide change, so it requires an RFC issue and other processes to be followed (see https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes), but no objections were raised
- https://github.com/zephyrproject-rtos/zephyr/pull/54525 should be reverted if this is merged
- prior discussion in https://github.com/zephyrproject-rtos/zephyr/pull/21100#pullrequestreview-325259287 
- discussion around to start, requiring 0 return value and reserve other values for future use (https://github.com/zephyrproject-rtos/zephyr/pull/54628#issuecomment-1431921596)
- request to verify consequences of not changing the return type and document migration steps in release notes

keith-packard committed:
samples/drivers/display: Eliminate control flow in macro

Remove RETURN_FROM_MAIN macro which performed control flow, violating
compliance tests. Replace with explicit use of posix_exit_main or return as
appropriate.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
> Arch WG:
> 
>     * This is a treewide change, so it requires an RFC issue and other processes to be followed (see https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes), but no objections were raised

#56504 

>     * [compiler/clang: ignore main-return-type warning for clang #54525](https://github.com/zephyrproject-rtos/zephyr/pull/54525) should be reverted if this is merged

These flags are removed as a part of this PR.

>     * request to verify consequences of not changing the return type and document migration steps in release notes

The only consequence of not changing that I'm aware of is that we will not be able to remove the `-ffreestanding` compiler flag when using clang as that has an un-skippable error for a non-int main return type (even if `-Wno-main-return-type` is specified). Consequences of the `-ffreestanding` compiler flag are discussed in PR #54638

keith-packard commented:
(I've rebased and fixed the conflicts today)

keith-packard commented:
> If we merge this, we need to "revert" (and similar changes in other files for `cmake/compiler/arcmwdt/compiler_flags.cmake`):

This change is already part of this PR.

carlescufi commented:
Setting DNM since this requires agreement before merging.

carlescufi reviewed (dismissed):
Please add an entry in the major changes sections of the release notes in:
https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/release-notes-3.4.rst

ifyall commented on samples/drivers/display/src/main.c:
Per the documentation we are adding in https://github.com/zephyrproject-rtos/zephyr/pull/54628/commits/1b765d022204ae6fa99835a19ffd54facc5fbe5f, a non-zero return type is not allowed at this time. This file doesn't adhere to that. I won't comment on the posix return case, as I am not familiar enough with the standard. However, should the non-posix flow return 0 in compliance with the doc? If not, then we should have comments in this file indicating why we are allowing the exception.

keith-packard replied:
oops! All fixed.

keith-packard committed:
doc/releases: Note "main" return type change for 3.4

Add a paragraph explaining the change from void main(void) to int
main(void) and the requirement for main functions to return zero.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
> Please add an entry in the major changes sections of the release notes in: https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/release-notes-3.4.rst

Done.

stephanosio reviewed (dismissed):
LGTM

stephanosio commented:
Needs a rebase.

keith-packard commented:
> Needs a rebase.

Yup, all done. No significant conflicts found.

keith-packard committed:
tests/fpu_sharing: Increase main stack size on riscv64

Looks like switching the main return value to int means that stack
frame persists and increases stack usage by a few bytes. Increase the
main stack size to avoid overflows.

Signed-off-by: Keith Packard <keithp@keithp.com>

stephanosio commented:
Will merge once approved in the TSC meeting today.

nordicjm reviewed (commented):
@aescolar seems like babblesim will need updating:
```
/*
 * This function will be called (in SW context) as the Zephyr application main
 */
typedef void (*bst_test_main_t)(void);
```
And various uses in tests:
```
./tests/bsim/bluetooth/host/privacy/peripheral/src/main.c:void main(void)
./tests/bsim/bluetooth/host/privacy/device/src/test_undirected_main.c:void main(void)
./tests/bsim/bluetooth/host/privacy/central/src/main.c:void main(void)
./tests/bsim/bluetooth/host/gatt/settings/src/main.c:void main(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_syncer.c:static void main_per_adv_syncer(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_syncer.c:static void main_per_adv_conn_syncer(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_syncer.c:static void main_per_adv_conn_privacy_syncer(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_syncer.c:static void main_per_adv_long_data_syncer(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_advertiser.c:static void main_per_adv_advertiser(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_advertiser.c:static void main_per_adv_conn_advertiser(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_advertiser.c:static void main_per_adv_conn_privacy_advertiser(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_advertiser.c:static void main_per_adv_long_data_advertiser(void)
./tests/bsim/bluetooth/host/adv/encrypted/css_sample_data/src/main.c:void main(void)
```

nordicjm reviewed (changes requested):
Following seem to have been missed in conversion:
```
./tests/drivers/build_all/mfd/src/main.c
./samples/subsys/logging/ble_backend/src/main.c
./samples/sensor/proximity_polling/src/main.c
./samples/fuel_gauge/max17048/src/main.c
./samples/drivers/virtualization/ivshmem/doorbell/src/ivshmem.c
./samples/drivers/smbus/src/main.c
./samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c
./samples/bluetooth/bthome_sensor_template/src/main.c
```
Documentation in ./arch/posix/include/posix_cheats.h should also be updated around this part:
```
#if defined(__cplusplus) && defined(CONFIG_CPP_MAIN)
/* To be able to define main() in C++ code we need to have its prototype
 * defined somewhere visibly. Otherwise name mangling will prevent the linker
 * from finding it. Zephyr assumes a void main(void) prototype and therefore
 * this will be the prototype after renaming:
 */
extern "C" int _posix_zephyr_main(void);
#endif
```

stephanosio commented:
Removing DNM since the proposal in #56504 has been approved by the TSC.

keith-packard committed:
tests/bsim: Switch main return type from void to int

These changes were automated using coccinelle with the following
script:

@@
@@
- void
+ int
main(...) {
    ...
-	return;
+	return 0;
    ...
}

Signed-off-by: Keith Packard <keithp@keithp.com>

nordicjm reviewed (changes requested):
Following are missing changes:
```
./tests/drivers/build_all/mfd/src/main.c:void main(void)
./samples/subsys/logging/ble_backend/src/main.c:void main(void)
./samples/sensor/proximity_polling/src/main.c:void main(void)
./samples/hello_world/src/main.c:void main(void)
./samples/fuel_gauge/max17048/src/main.c:void main(void)
./samples/drivers/virtualization/ivshmem/doorbell/src/ivshmem.c:void main(void)
./samples/drivers/smbus/src/main.c:void main(void)
./samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c:void main(void)
./samples/bluetooth/bthome_sensor_template/src/main.c:void main(void)
```

carlescufi commented:
@keith-packard there are no TSC objections to merging this. Could you rebase and address the missing ones that @nordicjm found? Thanks!

keith-packard committed:
samples, tests: Switch main return type from void to int

This applies the coccinelle script to another set of files:

   samples/bluetooth/bthome_sensor_template/src/main.c
   samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c
   samples/drivers/smbus/src/main.c
   samples/drivers/virtualization/ivshmem/doorbell/src/ivshmem.c
   samples/fuel_gauge/max17048/src/main.c
   samples/hello_world/src/main.c
   samples/sensor/proximity_polling/src/main.c
   samples/subsys/logging/ble_backend/src/main.c
   tests/drivers/build_all/mfd/src/main.c

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
> @keith-packard there are no TSC objections to merging this. Could you rebase and address the missing ones that @nordicjm found? Thanks!

rebased, repaired and re-pushed.

nordicjm reviewed (approved):
This will need to be removed but it can be done in a later commit, no point holding this up anymore:
```
config CPP_MAIN
        bool "C++ main() function definition"
        help
          This option instructs the Zephyr kernel to call the 'int main(void)'
          instead of the 'void main(void)', which is the default main() type
          for Zephyr.

          C++ does not allow the main() to be defined with 'void' return type,
          and any applications defining its main() in a C++ source file must
          enable this option.
```
CC @stephanosio 
And ./arch/posix/include/posix_cheats.h too

keith-packard commented:
I just realized that we can add `-Wmain` to the compiler flags if we want to catch applications using the wrong type for main. This works with gcc even with `-ffreestanding`, but doesn't appear to work with clang with that flag.
